spyrmsd/parallel.py

```diff
@@ -0,0 +1,127 @@
+    cache: bool = True,
+    num_workers: Optional[int] = None,
+    timeout: Optional[float] = None,
+    chunksize: int = 1,
+) -> List[float]:
+    """
+    Compute RMSD between two molecules with a timeout.
+    Parameters
+    ----------
+    molrefs: Union[molecule.Molecule, List[molecule.Molecule]]
+        Reference molecule
+    mols: Union[molecule.Molecule, List[molecule.Molecule]]
+        Molecules to compare to reference molecule
+    num_workers: int
+        Amount of processor to use for the parallel calculations
+    symmetry: bool, optional
+        Symmetry-corrected RMSD (using graph isomorphism)
+    center: bool, optional
+        Center molecules at origin
+    minimize: bool, optional
+        Minimised RMSD (using the quaternion polynomial method)
+    strip: bool, optional
+        Strip hydrogen atoms
+    timeout: float, optional
+        After how many seconds to stop the RMSD calculations
+    chunksize: int, optional
+        How many molecules to handle per process
+    Returns
+    -------
+    List[float]
+        RMSDs
+    """
+
+    # Ensure the num_workers is less or equal than the max number of CPUs.
+    # Silencing MyPy since os.cpu_count() can return None
+    if num_workers is None:
+        num_workers = os.cpu_count()
+    num_workers = min(num_workers, os.cpu_count())  # type: ignore[type-var]
+
+    if chunksize > 1 and timeout is not None:
+        warnings.warn("When using the timeout feature, a chunksize of 1 is required")
+        ## We could also raise an error here
+        chunksize = 1
+
+    # Cast the molecules to lists if they aren't already
+    if not isinstance(molrefs, list):
+        molrefs = [molrefs]
+    if not isinstance(mols, list):
+        mols = [mols]
+
+    # Match the length of the molref
+    if len(molrefs) == 1 and len(molrefs) < len(mols):
+        molrefs = molrefs * len(mols)
+
+    # Ensure molrefs and mols have the same len
+    if not len(molrefs) == len(mols):
+        raise ValueError("The 'mols' and 'molrefs' lists have different lengths.")
+
+    outputList = []
+
+    timeoutCounter = 0
+    errorCounter = 0
+
+    with ProcessPool(max_workers=num_workers) as pool:
+        rsmd_partial = partial(
+            rmsdwrapper,
+            symmetry=symmetry,
+            center=center,
+            minimize=minimize,
+            strip=strip,
+            cache=cache,
+        )
+
+        future = pool.map(
+            rsmd_partial, molrefs, mols, timeout=timeout, chunksize=chunksize
+        )
+        iterator = future.result()
+        idx = 0
+        while True:
+            try:
+                result = next(iterator)
+                outputList.append(result[0])
+            except StopIteration:
+                break
+            except TimeoutError:
+                timeoutCounter += 1
+                outputList += [np.nan] * chunksize
+            except Exception:
+                errorCounter += 1
+                outputList.append(np.nan)
```

Comment:
If we are iterating in chunks, exceptions are still for a single result? I'm not sure I understand why `TimeoutError` adds `chunksize` `np.nan`s while `Exception` only adds one, if they are working with the same iterator.

I probably need to play around with this a bit more to wrap my head around it.

Reply:
So this indeed is confusing, and one of the problems I was facing...

Here is an explanation of the author of `pebble`: https://github.com/noxdafox/pebble/issues/132#issuecomment-2105267462

Basically a "regular" error will only affect that specific entry, while a timeout error will make it so the whole chunk is lost. Because of that, I thought that this approach (appending np.nan with a "regular" error, while adding [np.nan]*chunksize) would fix the problem. However in some artificial test cases I saw inconsistent results, where the final output length didn't even match the length of the input list anymore. That's why I put the restriction of the timeout and the chunksize, to ensure that in production no unexpected errors like those can occur.

Reply:
> That's why I put the restriction of the timeout and the chunksize, to ensure that in production no unexpected errors like those can occur.

Apologies, I forgot about this. Could you please add an explanation (as comment) of why `if chunksize > 1 and timeout is not None:` is there, to remind our future selves? You could also add that this check is why you are using `np.nan` for timeouts and `[np.nan] * chunksize` otherwise.